### PR TITLE
Add a protoc version check into the build helpers.

### PR DESCRIPTION
**Pull request by thomasvl**
This doesn't help end developers using the plugin, but it atleast helps
validate that the generated files checked in are being generated with a current
enough protoc for their JSON support to be correct.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -210,6 +210,18 @@ build ${PROTOC_GEN_SWIFT}:
...
+# This will get run by any other rule that tries to use the plugin.  This hacks
+# in a check during the library build to ensure that the protoc on the local
+# system is one we expect. This was inspired by the findings that lead to
+#   https://github.com/apple/swift-protobuf/issues/111
+# This likely will need follow up to eventually do a pattern match to ensure it
+# is 3.1 or higher.
+${PROTOC_GEN_SWIFT}: build
+	@if [ "$(shell ${PROTOC} --version)" != "libprotoc 3.1.0" ]; then \
+	  echo "WARNING: Unexpected version of protoc: $(shell ${PROTOC} --version)"; \
+	  echo "WARNING: The JSON support in generate files may not be correct."; \
```
Comment (allevato) [inline Makefile]: s/generate/generated/

File: Makefile [hunk 2]
```diff
@@ -210,6 +210,18 @@ build ${PROTOC_GEN_SWIFT}:
...
 	${SWIFT} build
 
+# This will get run by any other rule that tries to use the plugin.  This hacks
+# in a check during the library build to ensure that the protoc on the local
+# system is one we expect. This was inspired by the findings that lead to
+#   https://github.com/apple/swift-protobuf/issues/111
+# This likely will need follow up to eventually do a pattern match to ensure it
+# is 3.1 or higher.
+${PROTOC_GEN_SWIFT}: build
+	@if [ "$(shell ${PROTOC} --version)" != "libprotoc 3.1.0" ]; then \
```
Comment (allevato) [inline Makefile]: Does this version number change infrequently enough that it's safe to punt on a more general pattern match until later?

Comment (thomasvl) [inline Makefile]: Not sure.  There were a few months between the 3.0.0 release and 3.1.0 releases.  There were some 3.0.x releases, but that number didn't always make it to the master branch.

The other catch, is I'm think the only shell we can rely on make spawning in /bin/sh, and test (`[`) doesn't have a regex matcher either.  So when this needs beefing up, it make get ugly and be yet another reason to move from a Makefile to a script for wrapping the build/test/regenerate process.

File: Makefile (hunk 1 above)
Comment (thomasvl) [inline Makefile]: Done (amended commit to include fix)

File: Makefile (hunk 2 above)
Comment (allevato) [inline Makefile]: Changing the != check to a grep for a few different version number patterns might not be that bad? I'll leave it up to you whether you want to try it now or file an issue and clean it up later.

Comment (tbkka) [inline Makefile]: grep returns success if nothing was selected, so you can use it with an 'if' in /bin/sh, using constructs like this:
```
     if ${PROTOC} --version | grep pattern ; then ...
```

**Pull request merged**